Fix two scenarios of RDataFrame JIT and explicit multithreading

vepadulano opened the pull request:
Re-opening [this PR](https://github.com/root-project/root/pull/15578) to check if the clad failures were real.

vepadulano committed:
[df] Add test for JIT with explicit multithreading

Test a program with multiple threads managed explicitly by the user and each thread runs an RDataFrame with JITting. In particular, this test probes the following situation:

* Threads start building the computation graph, code to JIT goes in a global accumulating string.
* Thread 1 reaches the RLoopManager::Jit method. Checks the global string, it is not empty.
* Thread 2 arrives as well, checks the string, it is not empty.
* Thread 1 now takes a write lock on the string, moving it into the local function stack, thus leaving it empty.
* Thread 2 wants to move the string as well, but just moves an empty string.
* Thread 2 is faster than Thread 1 at passing its (empty) string to the `InterpreterCalc` function.
* The `InterpreterCalc` function receives an empty string, thus will raise an exception since `gInterpreter->Calc("")` returns a `kDangerous` interpreter error code.

vepadulano committed:
[df] Ensure there is code to JIT before calling the interpreter

In an explicit multithreading scenario inside the `RLoopManager::Jit` method, a thread might move the global string with the code to JIT into its own function stack while another thread might still believe that string is not empty. This would make the other thread call `InterpreterCalc` with an empty string which would eventually lead to an exception. Check for the string emptiness twice, once at the beginning of the function, then again after taking the write lock so that if another thread already moved the string then the current thread will not continue with JITting.

pcanal commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -805,18 +805,25 @@ void RLoopManager::CleanUpTask(TTreeReader *r, unsigned int slot)
 void RLoopManager::Jit()
 {
-   {
+   auto noCodeToJit = []() {
```
The lock guard is redundant in the second invocation and thus could be extracted/used only for the first invocation. (Note that a thread asking for the write lock while holding the read lock is fully supported - read locks are suspended during the request for the write locks).

pcanal commented on tree/dataframe/test/dataframe_concurrency.cxx:
```diff
@@ -234,5 +243,32 @@ TEST(RDFConcurrency, ParallelRDFCachesEnableImplicitMT)
+   std::vector<int> results(25);
+   auto do_work = [&results](int slot) {
+      for (int i = slot * 5; i < (slot * 5 + 5); i++) {
+         results[i] = ROOT::RDataFrame{1}.Define("x", std::to_string(i)).Sum<int>("x").GetValue();
```
This test currently has random behavior (this is actually a good thing and we should keep as is but might want to add more tests) and thus in the extreme can spuriously succeed for example if the system choose to serialize all the threads.  We should probably add another version of the test that insure the ordering we are looking for (by using some sort of barriers, for example 'spin locks' over an atomic).

vepadulano replied:
I do not believe that is possible with the existing RDataFrame API. Since there is no way to call `RLoopManager::Jit`, there is no way for me to synchronize the jitting stages of the two RDFs. The only thing I can use as a synchronization trigger is the trigger of the event loop, but that means doing everything, calling `RLoopManager::Jit`, calling `RDFInternal::InterpreterCalc`. At that point, there is no way the second thread will have had time to reach the Jit function and take an empty string because it will have been blocked by the first thread's event loop and thus when it reaches the Jit function it just sees the string already empty and early returns

pcanal replied:
> At that point, there is no way the second thread will have had time to reach the Jit function and take an empty string because it will have been blocked by the first thread's event loop and thus when it reaches the Jit function it just sees the string already empty and early returns

You indeed right.

pcanal commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -813,10 +814,15 @@ void RLoopManager::Jit()
-      R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
-      return std::move(GetCodeToJit());
-   }();
+   R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
```
Consider the following code:
```
   R__READ_LOCKGUARD(ROOT::gCoreMutex);
   if (GetCodeToJit().empty()) {
      R__LOG_INFO(RDFLogChannel()) << "Nothing to jit and execute.";
      return;
   }

   R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
   // Check again if another thread has already cleared the global string
   // with the code to JIT. Without this check, we could end up calling
   // InterpreterCalc with an empty string, which would raise an exception.
   if (GetCodeToJit().empty()) {
      R__LOG_INFO(RDFLogChannel()) << "Nothing to jit and execute.";
      return;
   }
   const std::string code = std::move(GetCodeToJit());
```

vepadulano replied:
Just for my own understanding, your suggestion is almost equal to the current code, except that the first `R__READ_LOCKGUARD` is not enclosed in a scope but is at the "global" function scope. Why is it better to create a read lock first, then create a write lock in the same scope? The situation that should occur is that multiple threads may reach the read lock part, check for the emptiness of the string, then move to the write lock part where only one thread at a time can continue. But I don't understand how writing the read lock at function scope is different than writing it within an inner scope

pcanal replied:
The difference is that with the inner scope the operation are:

1. Take Read Lock
2. Give away Read lock
3. Take Write Lock

While without it, it is

1. Take Read Lock
2. Switch To Write Lock

So it reduces the lock churn, especially if there is no contention.

vepadulano replied:
Ah, cool then, thanks!

vepadulano commented:
This PR is related to this issue https://github.com/root-project/root/issues/15080

vepadulano committed:
[df] Add another test for RDF and explicit multithreading

For the following scenario:

* Multiple threads, multiple different RDF computation graphs (one per thread).
* Some threads run a valid, working computation graph, others have some
  problems.
* Two classes of problems are used. The first is when the declaration
  of some code to JIT incurs in an error in the interpreter. The second is when
  there is an exception at runtime.
* All problems are hidden explicitly by a try-catch scope.
* Threads running a working computation graph should not be impacted by the
  threads that are running a flawed computation graph.

vepadulano commented:
@pcanal A second test was added for a different type of scenario (see the commit description). On my local tests, the changes in this PR already cover also the second scenario.

hageboeck commented on tree/dataframe/test/dataframe_concurrency.cxx:
```diff
@@ -267,3 +277,32 @@ TEST(RDFConcurrency, HistoNSparseDInMT)
+   for (auto &&t : threads)
+      t.join();
+
+   expect_vec_eq(results, expected);
```
Couldn't this simply be:
```
EXPECT_TRUE(results == expected);
```
?

vepadulano replied:
gtest doesn't support equality of collections

hageboeck replied:
That's why I was suggesting the C++ operator `bool operator==(vector, vector)`

hageboeck commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -869,21 +869,24 @@ void RLoopManager::CleanUpTask(TTreeReader *r, unsigned int slot)
+      R__LOG_INFO(RDFLogChannel()) << "Nothing to jit and execute.";
+      return;
    }
 
-   std::string codeToDeclare, code;
-   {
-      R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
-      codeToDeclare.swap(GetCodeToDeclare());
-      code.swap(GetCodeToJit());
-   };
+   R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
```
Can this deadlock?

Imagine two threads manage to get to the read lock, and both decide that it's time to JIT, none of them can acquire the write lock because both hold the read lock.

vepadulano replied:
All ROOT `*LOCKGUARD` macros are reentrant so the read lock gets automatically promoted to a write lock. Similar pattern can be found in another places in ROOT e.g. https://github.com/root-project/root/blob/58db9f5db157008365d75ce53139644b7c143684/core/metacling/src/TCling.cxx#L6402-L6418

hageboeck replied:
I understand that if a single thread wants to promote its lock, there is no problem. But if two threads hold the read lock, none of them can acquire the write lock.
So there needs to be a mechanism to avoid deadlocks by releasing the read lock. That might either be done in ROOT's LOCKGUARD or the function needs to actively release the lock when it realises that it can't make progress.

vepadulano replied:
The algorithms are written in https://github.com/root-project/root/blob/master/core/thread/src/TReentrantRWLock.cxx

hageboeck replied:
Indeed the implementation can do something that I didn't imagine: the read locks are temporarily released while all threads wait for the write lock, so the code indeed works.
If we were to do it again, I would not use the gCoreMutex, but that has nothing to do with this bugfix.

vepadulano replied:
See my latest reply, I strongly believe we cannot do without `gCoreMutex`, I am open to being proven otherwise.

hageboeck reviewed (commented):
I find the locking logic too complicated (and invasive) for what RDF needs. What would you think of this:
- A `std:mutex` is added in the anonymous namespace next to the strings that need to be JITted.
- When *anything* happens to these strings, we `std::scoped_lock` that mutex.

Benefits:
- No locks on `gCoreMutex`. Why does RDF have to interfere with this mutex?
- No double checking if there's strings to JIT. With a scoped lock, there either is or isn't, and other threads can't interfere.
- C++ "native" logic, so easy to understand.

I also don't see that we would lose speed. If strings need to be JITted, the lock is only held for the time that we need to check and move, which is shorter than what this PR proposes. If new code needs to be enqueued, the lock is held for the same amount of time as before.

vepadulano commented:
> I find the locking logic too complicated (and invasive) for what RDF needs. What would you think of this:
> 
>     * A `std:mutex` is added in the anonymous namespace next to the strings that need to be JITted.
> 
>     * When _anything_ happens to these strings, we `std::scoped_lock` that mutex.
> 
> 
> Benefits:
> 
>     * No locks on `gCoreMutex`. Why does RDF have to interfere with this mutex?
> 
>     * No double checking if there's strings to JIT. With a scoped lock, there either is or isn't, and other threads can't interfere.
> 
>     * C++ "native" logic, so easy to understand.
> 
> 
> I also don't see that we would lose speed. If strings need to be JITted, the lock is only held for the time that we need to check and move, which is shorter than what this PR proposes. If new code needs to be enqueued, the lock is held for the same amount of time as before.

I see your reasoning, and I could agree philosophically speaking, but:

* The use of `R_*_LOCKGUARD` is in RDF since its inception. We can in principle think about moving away from it, but I would consider way beyond scope for this PR which *is not* revolutionizing the threading strategy in RDF, merely tightening its security with two concrete tests to accompany the changes
* As a starting food for thought on the topic of "can we move away from `R_*_LOCKGUARD` macros in RDF?". My first answer would be a strong no. The reason is that RDF uses heavily the interpreter. Moving away from `R_*_LOCKGUARD` would immediately expose the entire codebase to thread unsafety issues with the rest of ROOT as well as any downstream application that may use JITting outside of RDF in one or more threads.

vepadulano commented:
Just to add to my previous statement: since we control both the RDataFrame as well as the Core, Cling libraries, we could come up with a solution that delegates the use of `R_*_LOCKGUARD` macros safely only in Core and Cling, managing the RDF-only parts with standard C++ mutex and lock. I still think this is way beyond scope for this PR and I question if it's worth pursuing at all.

hageboeck reviewed (approved):
OK, let's do it like this:
Merge the fix and the tests proposed here. Then, in its own time, we can check if we want to have an RDF-local mutex.
